h2/src/main/org/h2/mvstore/tx/TransactionMap.java

```diff
@@ -728,16 +624,11 @@ protected TMIterator(TransactionMap<K,?> transactionMap, K from, K to, boolean i
             MVMap<K, VersionedValue> map = transactionMap.map;
             BitSet committingTransactions;
             MVMap.RootReference mapRootReference;
-            Page undoRoot;
             do {
                 committingTransactions = store.committingTransactions.get();
-                undoRoot = store.undoLog.getRootPage();
                 mapRootReference = map.getRoot();
-            } while (committingTransactions != store.committingTransactions.get()
-                    || undoRoot != store.undoLog.getRootPage());
+            } while (committingTransactions != store.committingTransactions.get());
```

Comment:
May be relationship between `map.getRoot()` and `store.committingTransactions.get()` should be briefly described? This loop is really executed more than once in some tests, so it's probably needed, but it's not very oblivious that `TransactionStore.commit()` changes them.

Also it looks a bit strange that similar loop in `sizeAsLong()` checks also `mapRootReference` in its condition. What the reason of this difference?

Reply:
Put some comments around it. There is 3-4 occurrences of similar loops scattered around, but I haven't figured out yet how to re-factor the code nicely (without creation of extra objects).